dbms/src/Compression/CompressionCodecDoubleDelta.cpp

```diff
@@ -0,0 +1,320 @@
+            if (double_delta > -63 && double_delta < 64)
+            {
+                writer.writeBits(2, 0b10);
+                writer.writeBits(1, sign);
+                writer.writeBits(6, abs_value);
+            }
+            else if (double_delta > -255 && double_delta < 256)
+            {
+                writer.writeBits(3, 0b110);
+                writer.writeBits(1, sign);
+                writer.writeBits(8, abs_value);
+            }
+            else if (double_delta > -2047 && double_delta < 2048)
+            {
+                writer.writeBits(4, 0b1110);
+                writer.writeBits(1, sign);
+                writer.writeBits(11, abs_value);
+            }
+            else if (double_delta > std::numeric_limits<Int32>::min() && double_delta < std::numeric_limits<Int32>::max())
+            {
+                writer.writeBits(5, 0b11110);
+                writer.writeBits(1, sign);
+                writer.writeBits(31, abs_value);
+            }
+            else
+            {
+                writer.writeBits(5, 0b11111);
+                writer.writeBits(1, sign);
+                writer.writeBits(63, abs_value);
+            }
+        }
+    }
+
+    writer.flush();
+
+    return sizeof(items_count) + sizeof(prev_value) + sizeof(prev_delta) + buffer.count();
+}
+
+template <typename T, typename DeltaType>
+void decompressDataForType(const char * source, UInt32 source_size, char * dest)
+{
+    const char * source_end = source + source_size;
+
+    const UInt32 items_count = unalignedLoad<UInt32>(source);
+    source += sizeof(items_count);
+
+    T prev_value{};
+    DeltaType prev_delta{};
+
+    if (source < source_end)
+    {
+        prev_value = unalignedLoad<T>(source);
+        unalignedStore(dest, prev_value);
+
+        source += sizeof(prev_value);
+        dest += sizeof(prev_value);
+    }
+
+    if (source < source_end)
+    {
+        prev_delta = unalignedLoad<DeltaType>(source);
+        prev_value = static_cast<DeltaType>(prev_value) + prev_delta;
+        unalignedStore(dest, prev_value);
+
+        source += sizeof(prev_delta);
+        dest += sizeof(prev_value);
+    }
+
+    ReadBufferFromMemory buffer(source, source_size - sizeof(prev_value) - sizeof(prev_delta) - sizeof(items_count));
+    BitReader reader(buffer);
+
+    // since data is tightly packed, up to 1 bit per value, and last byte is padded with zeroes,
+    // we have to keep track of items to avoid reading more that there is.
+    for (UInt32 items_read = 2; items_read < items_count && !reader.eof(); ++items_read)
+    {
+        DeltaType double_delta = 0;
+        if (reader.readBit() == 0)
+        {
+            double_delta = 0;
+        }
+        else
+        {
+            // first bit is 1
+            const UInt8 data_sizes[] = {6, 8, 11, 31, 63};
+            UInt8 i = 0;
+            for (; i < sizeof(data_sizes) - 1; ++i)
```

Comment:
It looks like a slow solution.

Reply:
'Looks' is a funny way to judge about performance ;) 
Anyway, what would you suggest me to do? Unroll a loop, construct a lookup-table? Anything else?

Reply:
I think we could implement countLeadingZeroes() and countLeadingOnes() for BitReader and try to optimize them.

Reply:
Am I reading this correct: no modifications to this part are required right now ?

Reply:
I'd prefer you extract this cycle into BitReader function for now and give it a good name.

Reply:
Moving code around wouldn't make it any faster.

Reply:
Well, I ran extra mile to optimize `BitReader` and `BitWriter` - now they have `UInt64`-buffers instead of `UInt8`, that means 8 times less calls to underlying buffer for reading or writing more data. IMO this should be performant enough now.